### PR TITLE
csf-v3-types rule

### PR DESCRIPTION
**Pull request by ahuth**
This PR adds the `csf-v3-types` rule. 

Per [@storybook/testing-react's documentation](https://github.com/storybookjs/testing-react/blob/eecb4767174ad3eb009209828c85ef8926842468/README.md?plain=1#L245) we need explicit types for `composeStories` to infer each story's args properly. This lint rule enforces that in the most common cases. As noted in comments, there are other cases that aren't worth the trouble to detect.

### CODE REVIEW

File: packages/eslint-plugin-stories/src/rules/__tests__/csf-v3-types.test.ts [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+      code: `
+        export const Secondary: StoryObj<Args> = {
+          ...Primary,
+          render: (args) => <Button {...args} />,
+        };
+      `,
+      filename: 'src/components/Button/Button.stories.tsx',
+    },
+    {
+      // Object story extending another one no render function.
```
Comment (ahuth) [inline packages/eslint-plugin-stories/src/rules/__tests__/csf-v3-types.test.ts:67]:
```suggestion
      // Object story extending another one with no render function.
```

File: packages/eslint-plugin-stories/src/rules/__tests__/csf-v3-types.test.ts [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+      // Bail out early and don't analyze these. Possible false negative, but it's not worth trying
+      // to figure out whether the properties brought in by the spread object would require us to
+      // type this thing or not.
```
Comment (ahuth) [inline packages/eslint-plugin-stories/src/rules/__tests__/csf-v3-types.test.ts:70]: Basically, for cases like

```js
export const Secondary = {
  ...Primary,
};
```

whether it needs an explicit type or not really depends on what's going on in `Primary`.

But I didn't want to try to figure out how to look at that node. I figure that we don't need to catch 100% of cases for this rule to be useful, and it's only worth jumping through so many hoops.

File: packages/eslint-plugin-stories/src/rules/csf-v3-types.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+ *   ```
+ *
+ * But these other cases aren't worth jumping through hoops to find.
+ */
+const rule: TSESLint.RuleModule<'specifyType'> = {
+  meta: {
+    messages: {
+      specifyType:
+        'Specify a type of `StoryObj<Args>` for object stories, so that `composeStories` and Storybook can infer the correct args',
+    },
```
Comment (ahuth) [inline packages/eslint-plugin-stories/src/rules/csf-v3-types.ts:39]: Using `TSESLint.RuleModule` is necessary to get the right TS types for the AST nodes. Well, that or some type casting.

What's weird about `TSESLint.RuleModule` is it doesn't let you do

```js
context.report({
  node,
  message: 'some message',
});
```

Instead it forces you (via the type system) to specify the message in `meta` and pass in the message id.

At runtime it works fine to pass in `message`, but I wonder if the folks at `@typescript-eslint` really wanted to enforce usage of `messageId` for some reason.

File: packages/eslint-plugin-stories/src/rules/csf-v3-types.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    const declarator = declaration.declarations[0];
+
+    // The "initial" expression is the value of the variable.
+    // i.e.
+    const expression = declarator.init;
+
+    // If the value is an object literal, assume its a story.
+    return expression?.type === 'ObjectExpression';
+  }
+  return false;
```
Comment (ahuth) [inline packages/eslint-plugin-stories/src/rules/csf-v3-types.ts:159]: This whole function exists just to cleanup the rule definition.

But I'm not going to pretend like it's the prettiest code I ever wrote. Let me know if you see ways to clean it up, or if it's not worth extracting in the first place.

File: packages/eslint-plugin-stories/src/index.ts [hunk 5]
```diff
@@ -11,6 +13,7 @@ const recommendedRules = {
 };
 
 const strictRules = {
+  '@chanzuckerberg/stories/csf-v3-types': 'error',
```
Comment (ahuth) [inline packages/eslint-plugin-stories/src/index.ts:16]: Maybe we should keep TS and JS rules separate, and make a `ts-strict` config?

File: packages/eslint-plugin-stories/src/rules/csf-v3-types.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+ *
+ * Only detects 2 cases:
+ *
+ * 1. Stories that are empty objects - `const Primary = {}`
+ * 2. Stories with a render function that takes args - `render: (args) => { ... }`
+ *
+ * Other cases exist that need types, but they're harder to detect statically. These 2 are likely
+ * the most common.
+ */
+const rule: TSESLint.RuleModule<'specifyType'> = {
```
Comment (ahuth) [inline packages/eslint-plugin-stories/src/rules/csf-v3-types.ts:28]: Note the use of `TSESLint.RuleModule` instead of the standard `Rule.RuleModule`.

Necessary to get god `@typescript-eslint/parser` types. They're compatible, but the `TSESLint` version enforces via the type system defining the rule in an oddly specific way.

For example, it forces us to define `meta`, and to specify failure messages via `messageId` (instead of directly passing a string to `report`).

TL;DR - all this is valid eslint rule config, but it still looks a bit different than other rules in this plugin.

File: packages/eslint-plugin-stories/src/rules/csf-v3-types.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+  create(context) {
+    if (!isStories(context.getFilename())) {
+      return {};
+    }
+
+    return {
+      // Examine any named exports. In a stories file, these are likely to be stories.
+      // i.e. `export const foo = {}`
+      ExportNamedDeclaration(node) {
+        const story = StoryObject.parse(node);
```
Comment (ahuth) [inline packages/eslint-plugin-stories/src/rules/csf-v3-types.ts:41]: Inspried by https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/ and https://www.oreilly.com/library/view/code-that-fits/9780137464302/

File: packages/eslint-plugin-stories/src/rules/csf-v3-types.ts [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+ * this rules enforces adding a type annotation
+ *
+ *   const Primary: StoryObj<Args> = {
+ *     render: (args) => { ... },
+ *   }
+ *
+ * Only detects instances where we're sure a type annotation is needed. For example, when there's a
+ * render function defined that takes args (as in the examples above).
+ */
+const rule: TSESLint.RuleModule<keyof typeof failureMessages> = {
```
Comment (ahuth) [inline packages/eslint-plugin-stories/src/rules/csf-v3-types.ts:28]: Using `TSESLint.RuleModule` instead of the standard `Rule.RuleModule`.

Necessary to get good `@typescript-eslint/parser` types. They're compatible, but the `TSESLint` version enforces (via the type system) defining the rule in an oddly specific way.

For example, we have to define the optional `meta` property, and specify failure messages via `messageId` (instead of directly passing a message to `context.report`).

TL;DR - all this is valid eslint rule config, but it still looks a bit different than other rules in this plugin.
